### ISSUE TITLE
No slots on ALPS/COBALT

### ISSUE THREAD

**Issue by abouteiller** [openpmix/prrte#577]
`mpirun` cannot launch processes on ALCF Theta, because the RAS component cannot detect the allocation.

If processes are launched with aprun, the behavior is that each MPI process is in a singleton world.

## Background information

```
 4a43c39c89037f52b4e25927e58caf08f3707c33 opal/mca/hwloc/hwloc2/hwloc (hwloc-2.1.0rc2-53-g4a43c39c)
 4996218a7a13a0ec59c94ffd72ead6d95d6acb4e opal/mca/pmix/pmix4x/openpmix (v1.1.3-2366-g4996218a)
 9cf321a38d36900702fd9c0862f633f4981b740d prrte (dev-30542-g9cf321a38d)
```

### Please describe the system on which you are running

ALCF Theta

* Operating system/version: CrayPE
* Computer hardware: KNL
* Network type: Airies

-----------------------------

## Details of the problem

```
qsub -t 1:00:00 -n 8 -qdebug-cache-quad -A XYZ -I
Connecting to thetamom3 for interactive qsub...
Job routed to queue "debug-cache-quad".
Memory mode set to cache quad for queue debug-cache-quad
Wait for job 440524 to start...
Opening interactive session to 3824-3827,3832-3833,3838-3839
```

```
abouteil@thetamom3:/gpfs/mira-home/abouteil/ompi/pr7740> env | grep COBALT
COBALT_BLOCKNAME=3824-3827,3832-3833,3838-3839
COBALT_PARTNAME=3824-3827,3832-3833,3838-3839
COBALT_JOBID=440524
COBALT_PARTSIZE=8
COBALT_NODEFILE=/var/tmp/cobalt.440524
COBALT_JOBSIZE=8
COBALT_BLOCKSIZE=8
abouteil@thetamom3:/gpfs/mira-home/abouteil/ompi/pr7740> env | grep ALPS
CRAYPAT_ALPS_COMPONENT=/opt/cray/pe/perftools/7.0.5/sbin/pat_alps
PE_PRODUCT_LIST=CRAYPE_MIC-KNL:CRAY_RCA:CRAY_PMI:CRAY_LIBSCI:CRAYPE:INTEL:CRAY_XPMEM:CRAY_DMAPP:CRAY_UGNI:CRAY_UDREG:CRAY_ALPS:PERFTOOLS:CRAYPAT
CRAY_ALPS_POST_LINK_OPTS=-L/opt/cray/alps/6.6.43-6.0.7.1_5.51__ga796da32.ari/lib64
CRAY_ALPS_INCLUDE_OPTS=-I/opt/cray/alps/6.6.43-6.0.7.1_5.51__ga796da32.ari/include
abouteil@thetamom3:/gpfs/mira-home/abouteil/ompi/pr7740> env | grep BASIL
```

Failure to launch because RAS cannot find ALPS reservation:
```
build.master/bin/mpirun -N 1 -n 2  --mca rmaps_base_verbose 100  --debug --debug-daemons --debug-verbose 100  --display-devel-allocation  --mca ras_base_verbose 100 hostname
[thetamom3:52668] procdir: /tmp/prrte.thetamom3.32857/dvm.52668/0/0
[thetamom3:52668] jobdir: /tmp/prrte.thetamom3.32857/dvm.52668/0
[thetamom3:52668] job: /tmp/prrte.thetamom3.32857/dvm.52668
[thetamom3:52668] top: /tmp/prrte.thetamom3.32857
[thetamom3:52668] tmp: /tmp
[thetamom3:52668] sess_dir_cleanup: job session dir does not exist
[thetamom3:52668] sess_dir_cleanup: top session dir does not exist
[thetamom3:52668] procdir: /tmp/prrte.thetamom3.32857/dvm.52668/0/0
[thetamom3:52668] jobdir: /tmp/prrte.thetamom3.32857/dvm.52668/0
[thetamom3:52668] job: /tmp/prrte.thetamom3.32857/dvm.52668
[thetamom3:52668] top: /tmp/prrte.thetamom3.32857
[thetamom3:52668] tmp: /tmp
[thetamom3:52668] mca: base: components_register: registering framework ras components
[thetamom3:52668] mca: base: components_register: found loaded component simulator
[thetamom3:52668] mca: base: components_register: component simulator register function successful
[thetamom3:52668] mca: base: components_register: found loaded component alps
[thetamom3:52668] mca: base: components_register: component alps register function successful
[thetamom3:52668] mca: base: components_register: found loaded component slurm
[thetamom3:52668] mca: base: components_register: component slurm register function successful
[thetamom3:52668] mca: base: components_open: opening ras components
[thetamom3:52668] mca: base: components_open: found loaded component simulator
[thetamom3:52668] mca: base: components_open: found loaded component alps
[thetamom3:52668] mca: base: components_open: component alps open function successful
[thetamom3:52668] mca: base: components_open: found loaded component slurm
[thetamom3:52668] mca: base: components_open: component slurm open function successful
[thetamom3:52668] mca:base:select: Auto-selecting ras components
[thetamom3:52668] mca:base:select:(  ras) Querying component [simulator]
[thetamom3:52668] mca:base:select:(  ras) Querying component [alps]



# THIS IS PROBABLY THE ROOT CAUSE
[thetamom3:52668] ras:alps: NOT available for selection -- OMPI_ALPS_RESID or BASIL_RESERVATION_ID not set?




[thetamom3:52668] mca:base:select:(  ras) Querying component [slurm]
[thetamom3:52668] mca:base:select:(  ras) No component selected!
[thetamom3:52668] mca: base: components_register: registering framework rmaps components
[thetamom3:52668] mca: base: components_register: found loaded component mindist
[thetamom3:52668] mca: base: components_register: component mindist register function successful
[thetamom3:52668] mca: base: components_register: found loaded component ppr
[thetamom3:52668] mca: base: components_register: component ppr register function successful
[thetamom3:52668] mca: base: components_register: found loaded component rank_file
[thetamom3:52668] mca: base: components_register: component rank_file register function successful
[thetamom3:52668] mca: base: components_register: found loaded component resilient
[thetamom3:52668] mca: base: components_register: component resilient register function successful
[thetamom3:52668] mca: base: components_register: found loaded component round_robin
[thetamom3:52668] mca: base: components_register: component round_robin register function successful
[thetamom3:52668] mca: base: components_register: found loaded component seq
[thetamom3:52668] mca: base: components_register: component seq register function successful
[thetamom3:52668] [[14954,0],0] rmaps:base set policy with NULL device NONNULL
[thetamom3:52668] mca: base: components_open: opening rmaps components
[thetamom3:52668] mca: base: components_open: found loaded component mindist

[thetamom3:52668] mca: base: components_open: component resilient open function successful                                                                                                                  [8/1724]
[thetamom3:52668] mca: base: components_open: found loaded component round_robin
[thetamom3:52668] mca: base: components_open: component round_robin open function successful
[thetamom3:52668] mca: base: components_open: found loaded component seq
[thetamom3:52668] mca: base: components_open: component seq open function successful
[thetamom3:52668] mca:rmaps:select: checking available component mindist
[thetamom3:52668] mca:rmaps:select: Querying component [mindist]
[thetamom3:52668] mca:rmaps:select: checking available component ppr
[thetamom3:52668] mca:rmaps:select: Querying component [ppr]
[thetamom3:52668] mca:rmaps:select: checking available component rank_file
[thetamom3:52668] mca:rmaps:select: Querying component [rank_file]
[thetamom3:52668] mca:rmaps:select: checking available component resilient
[thetamom3:52668] mca:rmaps:select: Querying component [resilient]
[thetamom3:52668] mca:rmaps:select: checking available component round_robin
[thetamom3:52668] mca:rmaps:select: Querying component [round_robin]
[thetamom3:52668] mca:rmaps:select: checking available component seq
[thetamom3:52668] mca:rmaps:select: Querying component [seq]
[thetamom3:52668] [[14954,0],0]: Final mapper priorities
[thetamom3:52668]       Mapper: ppr Priority: 90
[thetamom3:52668]       Mapper: seq Priority: 60
[thetamom3:52668]       Mapper: resilient Priority: 40
[thetamom3:52668]       Mapper: mindist Priority: 20
[thetamom3:52668]       Mapper: round_robin Priority: 10
[thetamom3:52668]       Mapper: rank_file Priority: 0
[thetamom3:52668] [[14954,0],0] rmaps:base set policy with ppr:1:node device NULL
[thetamom3:52668] [[14954,0],0] rmaps:base policy ppr modifiers 1:node provided

======================   ALLOCATED NODES   ======================
=================================================================

======================   ALLOCATED NODES   ======================
        prte: slots=0 max_slots=0 slots_inuse=0 state=UP
        flags: DAEMON_LAUNCHED:LOCATION_VERIFIED
        thetamom3: slots=1 max_slots=0 slots_inuse=0 state=UP
        flags: NONE
=================================================================
[thetamom3:52668] mca:rmaps: mapping job [14954,1]
[thetamom3:52668] mca:rmaps: dynamic job [14954,1] will not inherit launch directives - parent [14954,0] is NON-TOOL
[thetamom3:52668] mca:rmaps: setting mapping policies for job [14954,1] nprocs 2
[thetamom3:52668] mca:rmaps[311] binding not given - using bycore
[thetamom3:52668] mca:rmaps:ppr: mapping job [14954,1] with ppr 1:node
[thetamom3:52668] mca:rmaps:ppr: job [14954,1] assigned policy BYNODE:NO_USE_LOCAL,NOOVERSUBSCRIBE
--------------------------------------------------------------------------
All nodes which are allocated for this job are already filled.
--------------------------------------------------------------------------
[thetamom3:52668] PRRTE ERROR:  in file ../../../../../../master/prrte/src/mca/rmaps/ppr/rmaps_ppr.c at line 254
[thetamom3:52668] [[14954,0],0] Releasing job data for [INVALID]
[thetamom3:52668] [[14954,0],0] prted_cmd: received exit cmd
[thetamom3:52668] [[14954,0],0] prted_cmd: all routes and children gone - exiting
[thetamom3:52668] [[14954,0],0] Releasing job data for [14954,0]
```

If the hosts are forced, one can launch PRTE and non-MPI processes
```
build.master/bin/mpirun -N 1 -n 2  --mca rmaps_base_verbose 100  --debug --debug-daemons --debug-verbose 100  --display-devel-allocation  --mca ras_base_ve
rbose 100 --host $COBALT_BLOCKNAME hostname
[thetamom3:00653] mca:base:select: Auto-selecting ras components
[thetamom3:00653] mca:base:select:(  ras) Querying component [simulator]
[thetamom3:00653] mca:base:select:(  ras) Querying component [alps]
[thetamom3:00653] ras:alps: NOT available for selection -- OMPI_ALPS_RESID or BASIL_RESERVATION_ID not set?
[thetamom3:00653] mca:base:select:(  ras) Querying component [slurm]
[thetamom3:00653] mca:base:select:(  ras) No component selected!

[thetamom3:00653] [[62142,0],0]: Final mapper priorities
[thetamom3:00653]       Mapper: ppr Priority: 90
[thetamom3:00653]       Mapper: seq Priority: 60
[thetamom3:00653]       Mapper: resilient Priority: 40
[thetamom3:00653]       Mapper: mindist Priority: 20
[thetamom3:00653]       Mapper: round_robin Priority: 10
[thetamom3:00653]       Mapper: rank_file Priority: 0
[thetamom3:00653] [[62142,0],0] rmaps:base set policy with ppr:1:node device NULL
[thetamom3:00653] [[62142,0],0] rmaps:base policy ppr modifiers 1:node provided

======================   ALLOCATED NODES   ======================
=================================================================

======================   ALLOCATED NODES   ======================
        3824-3827: slots=1 max_slots=0 slots_inuse=0 state=UP
        flags: SLOTS_GIVEN
        3832-3833: slots=1 max_slots=0 slots_inuse=0 state=UP
        flags: SLOTS_GIVEN
        3838-3839: slots=1 max_slots=0 slots_inuse=0 state=UP
        flags: SLOTS_GIVEN
=================================================================
[thetamom3:00653] plm:alps: final top-level argv:
[thetamom3:00653] plm:alps:     aprun -n 3 -N 1 -cc none -e PMI_NO_PREINITIALIZE=1 -e PMI_NO_FORK=1 -e OMPI_NO_USE_CRAY_PMI=1 -L 3824-3827,3832-3833,3838-3839 prted --debug --debug-daemons --prtemca ess_base_nspa
ce mpirun-thetamom3653 --prtemca ess_base_vpid 1 --prtemca ess_base_num_procs 4 --prtemca prrte_hnp_uri mpirun-thetamom3653.0;tcp://10.128.3.55,10.237.16.104:48384:14,16
[thetamom3:00653] plm:alps: Set prefix:/home/abouteil/ompi/pr7740/build.master

Daemon was launched on nid03824 - beginning to initialize
Daemon was launched on nid03825 - beginning to initialize
Daemon was launched on nid03826 - beginning to initialize


# (x3)
Daemon [[62142,0],1] checking in as pid 194498 on host nid03824
[thetamom3:00653] [[62142,0],0] prted_cmd: received pass_node_info
[thetamom3:00653] mca:rmaps: mapping job [62142,1]
[thetamom3:00653] mca:rmaps: dynamic job [62142,1] will not inherit launch directives - parent [62142,0] is NON-TOOL
[thetamom3:00653] mca:rmaps: setting mapping policies for job [62142,1] nprocs 2
[thetamom3:00653] mca:rmaps[311] binding not given - using bycore
[thetamom3:00653] mca:rmaps:ppr: mapping job [62142,1] with ppr 1:node
[thetamom3:00653] mca:rmaps:ppr: job [62142,1] assigned policy BYNODE:NO_USE_LOCAL,NOOVERSUBSCRIBE
[thetamom3:00653] AVAILABLE NODES FOR MAPPING:
[thetamom3:00653]     node: 3824-3827 daemon: 1
[thetamom3:00653]     node: 3832-3833 daemon: 2
[thetamom3:00653]     node: 3838-3839 daemon: 3
[thetamom3:00653] mca:rmaps: assigning locations for job [62142,1]
[thetamom3:00653] mca:rmaps:ppr: assigning locations for job [62142,1] with ppr 1:node policy BYNODE:NO_USE_LOCAL,NOOVERSUBSCRIBE
[thetamom3:00653] RANKING POLICY: SLOT
[thetamom3:00653] mca:rmaps:base: computing vpids by slot for job [62142,1]
[thetamom3:00653] mca:rmaps:base: assigning rank 0 to node 3824-3827
[thetamom3:00653] mca:rmaps:base: assigning rank 1 to node 3832-3833
[thetamom3:00653] [[62142,0],0] prted_cmd: received add_local_procs
[thetamom3:00653] [[62142,0],0] Releasing job data for [INVALID]
[thetamom3:00653] mca:rmaps: compute bindings for job [62142,1] with policy CORE:IF-SUPPORTED[1008]
[thetamom3:00653] mca:rmaps: computing bindings for job [62142,1]


# more outputs, looks like all works as intended
```

If launching an MPI proc, I get a MPI_INIT error in rte_init
```
--------------------------------------------------------------------------
It looks like MPI runtime init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during RTE init; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  local size
  --> Returned "Not found" (-13) instead of "Success" (0)
--------------------------------------------------------------------------
--------------------------------------------------------------------------
It looks like MPI_INIT failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during MPI_INIT; some of which are due to configuration or environment
problems.  This failure appears to be an internal failure; here's some
additional information (which may only be relevant to an Open MPI
developer):

  ompi_mpi_init: ompi_rte_init failed
  --> Returned "Not found" (-13) instead of "Success" (0)
--------------------------------------------------------------------------
*** An error occurred in MPI_Init_thread
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[3824-3827:194754] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```

If I `aprun`  the MPI processes, they all run in different `comm_world` with size=1 each.

**Comment by rhc54** [comment]
Well, this issue is really several things all jammed into one, so let's try to deal with them one at a time.

**Issue 1: no RAS component for Cobalt**
Yes, of course there isn't a RAS component for Cobalt. Nobody ever wrote one or requested one. Given that Cobalt is transitioning/merging with PBSPro, it is doubtful that (a)  it is worth the effort and (b) would happen in a timely manner given everything else that is being done.

**Issue 2: Local size not found**
I don't know why that would happen - it could be a problem in the odls/alps component. I don't have access to an ALPS machine, so someone who does would have to investigate. Note that things are working on other Cray machines (in MTT and CI), so this may be another symptom of the mixed Cobalt/Cray environment

**Issue 3: aprun results in every process thinking it is a singleton**
Yes, of course that would happen - we have repeatedly stated that OMPI master/v5 does _not_ support PMI2, which is what `aprun` uses. So there is no way you can launch an MPI job based on OMPI master using `aprun`.

**Comment by hppritcha** [comment]
@abouteiller I have adding a COBALT RAS to prrte on my to do list.  Hopefully I'll get to it before end of June.  Note there are different flavors of cobalt configs so we can't just go by what one sees on theta.

**Comment by hppritcha** [comment]
The ALPS odls is not working because your daemon processes are not part of the Cray ```PAGG``` container, because your just running on the mom node.

**Comment by rhc54** [comment]
@abouteiller Can you please give https://github.com/openpmix/prrte/pull/588 a try and see if it fixes the problem?

**Comment by rhc54** [comment]
@abouteiller The problem we may hit here is that it will think there is only one slot/node due to the way the hostnames are provided. Do you know if COBALT_PARTSIZE reflects the number of slots? It seems too low for the core count - I might have to do some additional fiddling to allow us to auto-discover the number of cores on the nodes.

**Comment by rhc54** [comment]
Oh BTW: you could fix the attempt:
```text
mpirun -N 1 -n 2  --mca rmaps_base_verbose 100  --debug --debug-daemons --debug-verbose 100  --display-devel-allocation  --mca ras_base_verbose 100 --host $COBALT_BLOCKNAME
```
for an MPI job by simply adding `--nolocal` to the cmd line. As @hppritcha noted, because you don't have that option on the cmd line, `mpirun` was attempting to run at least one proc on the login node, which isn't supported.

**Comment by rhc54** [comment]
@roblatham00 You might be interested in testing #588 out too, if you get a chance. As noted above, I think we still need a way to determine the number of allocated slots on each node, but I first would like to see if that PR solves the problem of detecting what was allocated.

**Comment by hppritcha** [comment]
Hmm... there's more going on here than I initially though.  I'll open a PR for a very small change which gets the ALPS RAS working with cobalt, but I'm still hitting an issue in the ALPS PLM.  It seems that somehow alps/apsched doesn't think we're in a reservation by the time we get to the execve operation.  I opened a ticket with ALCF support that should get routed to the COBALT crew.

**Comment by hppritcha** [comment]
fixed via #593
